core/src/main/java/org/apache/iceberg/avro/AvroMetrics.java

```diff
@@ -31,7 +42,112 @@ private AvroMetrics() {
 
   static Metrics fromWriter(DatumWriter<?> datumWriter, Schema schema, long numRecords,
                             MetricsConfig inputMetricsConfig) {
-    // TODO will populate in following PRs if datum writer is a MetricsAwareDatumWriter
-    return new Metrics(numRecords, null, null, null);
+    if (!(datumWriter instanceof MetricsAwareDatumWriter)) {
+      return new Metrics(numRecords, null, null, null, null);
+    }
+
+    MetricsAwareDatumWriter<?> metricsAwareDatumWriter = (MetricsAwareDatumWriter<?>) datumWriter;
+    MetricsConfig metricsConfig;
+    if (inputMetricsConfig == null) {
+      metricsConfig = MetricsConfig.getDefault();
+    } else {
+      metricsConfig = inputMetricsConfig;
+    }
+
+    Map<Integer, Long> valueCounts = new HashMap<>();
```

Comment:
Nit: we typically prefer `Maps.newHashMap()`, which will add null checking, and no put value should be null here.